Modules/_datetimemodule.c

```diff
@@ -2940,46 +2949,62 @@ datetime_date_fromtimestamp_capi(PyObject *cls, PyObject *args)
     return result;
 }
 
-/* Return new date from proleptic Gregorian ordinal.  Raises ValueError if
- * the ordinal is out of range.
- */
+
+/*[clinic input]
+@classmethod
+datetime.date.fromordinal
+
+    ordinal: int
+    /
+
+int -> date corresponding to a proleptic Gregorian ordinal.
+
+Raises ValueError if the ordinal is out of range.
+[clinic start generated code]*/
+
 static PyObject *
-date_fromordinal(PyObject *cls, PyObject *args)
+datetime_date_fromordinal_impl(PyTypeObject *type, int ordinal)
+/*[clinic end generated code: output=ea5cc69d86614a6b input=1d7158c082e677fd]*/
 {
     PyObject *result = NULL;
-    int ordinal;
-
-    if (PyArg_ParseTuple(args, "i:fromordinal", &ordinal)) {
-        int year;
-        int month;
-        int day;
+    int year;
+    int month;
+    int day;
 
-        if (ordinal < 1)
-            PyErr_SetString(PyExc_ValueError, "ordinal must be "
-                                              ">= 1");
-        else {
-            ord_to_ymd(ordinal, &year, &month, &day);
-            result = new_date_subclass_ex(year, month, day, cls);
-        }
+    if (ordinal < 1)
+        PyErr_SetString(PyExc_ValueError, "ordinal must be >= 1");
+    else {
+        ord_to_ymd(ordinal, &year, &month, &day);
+        result = new_date_subclass_ex(year, month, day, (PyObject *) type);
     }
     return result;
 }
 
-/* Return the new date from a string as generated by date.isoformat() */
+/*[clinic input]
+@classmethod
+datetime.date.fromisoformat
+
+    date_string: object
```

Comment:
We can use `date_string as dstr` here to avoid changing the C variable name.

```suggestion
    date_string as dstr: object
```

Reply:
Is that preferable? datetime.py also uses `date_string` as the variable name in `fromisoformat`.

Reply:
It won't be user-visible, it will only keep the internal C variable name `dstr`, to avoid unnecessary code churn.

Reply:
Addressed!